subcommands/targets/offline-update.go

```diff
@@ -65,25 +71,54 @@ func init() {
 		"Skip fetching Target Apps")
 	offlineUpdateCmd.Flags().BoolVarP(&ouAllowMultipleTargets, "allow-multiple-targets", "", false,
 		"Allow multiple targets to be stored in the same <dst> directory")
+	offlineUpdateCmd.Flags().StringVarP(&ouWave, "wave", "", "",
+		"Name of the wave Target content of which to fetch")
 }
 
 func doOfflineUpdate(cmd *cobra.Command, args []string) {
 	factory := viper.GetString("factory")
 	targetName := args[0]
 	dstDir := args[1]
 
-	if len(ouTag) == 0 {
-		subcommands.DieNotNil(errors.New("missing mandatory flag `--tag`"))
+	if len(ouTag) == 0 && len(ouWave) == 0 {
+		subcommands.DieNotNil(errors.New("one of the mandatory parameters `--tag` or `--wave` should be specified"))
+	}
+
+	if len(ouTag) > 0 && len(ouWave) > 0 {
+		subcommands.DieNotNil(errors.New("either `--tag` or `--wave` should be specified exclusively"))
 	}
 
-	fmt.Printf("Checking whether Target exists; target: %s, tag: %s, production: %v\n", targetName, ouTag, ouProd)
-	subcommands.DieNotNil(checkIfTargetExists(factory, targetName, ouTag, ouProd))
+	if len(ouWave) > 0 && ouProd {
+		subcommands.DieNotNil(errors.New("`the --prod` flag is redundant if the `--wave` parameter is specified"))
+	}
 
-	ti, err := getTargetInfo(factory, targetName)
+	var getTargetMeta func() (*tuf.FileMeta, error)
+	if len(ouWave) > 0 {
+		getTargetMeta = func() (*tuf.FileMeta, error) {
+			fmt.Printf("Getting Wave Target details; target: %s, wave: %s...\n", targetName, ouWave)
+			return getWaveTargetMeta(factory, targetName, ouWave)
+		}
+	} else if ouProd {
+		getTargetMeta = func() (*tuf.FileMeta, error) {
+			fmt.Printf("Getting production Target details; target: %s, tag: %s...\n", targetName, ouTag)
+			return getProdTargetMeta(factory, targetName, ouTag)
+		}
+	} else {
+		getTargetMeta = func() (*tuf.FileMeta, error) {
+			fmt.Printf("Getting CI Target details; target: %s, tag: %s...\n", targetName, ouTag)
+			return getCiTargetMeta(factory, targetName, ouTag)
+		}
+	}
+	// Get the wave/prod/CI specific target with the specified tag to check if it is really present, and
+	// get the specified target from the list of factory targets to obtain the "original" tag/branch that produced
+	// the target so we can find out the correct app bundle fetch URL.
+	ti, err := checkAndGetTargetInfo(getTargetMeta, func() (*tuf.FileMeta, error) {
+		return api.TargetGet(factory, targetName)
+	})
```

Comment:
IIUC you simply ignore the return of the `getTargetMeta` inside the `checkAndGetTargetInfo`, only processing the error.

So, just IMHO, the code flow would look more organically (easier to understand) if you do the following:
1. Rename all of `getWaveTargetMeta, getProdTargetMeta, getCiTargetMeta` into `checkXXXExists` returning only an error.
2. Check for the respective targets existance in that large `if..else if...else` clause above.
3. Keep the old vanilla `getTargetInfo` as is and use it here.

All in all, what I suggest is this:
```
	// Your big comment would better look here
	if len(ouWave) > 0 {
		fmt.Printf("Getting Wave Target details; target: %s, wave: %s...\n", targetName, ouWave)
		subcommands.DieNotNil(checkIfWaveTargetExists(factory, targetName, ouWave), "Failed to get a wave target with given parameters")
		// You might preserve your current signature if you wish, it would just make a code a bit longer... but I'm fine with that.
		// _, err := getWaveTargetMeta(factory, targetName, ouWave)
		// subcommands.DieNotNil(err, "Failed to get a wave target with given parameters")
	} else if ouProd {
		fmt.Printf("Getting production Target details; target: %s, tag: %s...\n", targetName, ouTag)
		subcommands.DieNotNil(checkIfProdTargetExists(factory, targetName, ouTag), "Failed to get a production target with given parameters")
	} else {
		fmt.Printf("Getting CI Target details; target: %s, tag: %s...\n", targetName, ouTag)
		subcommands.DieNotNil(checkIfCiTargetExists(factory, targetName, ouTag), "Failed to get a CI target with given parameters")
	}
	ti, err := getTargetInfo(factory, targetName)
```

In addition to the above it looks like in case of CI targets you're trying to fetch CI targets twice, right?
TBH I would optimize that use case. It is OK to duplicate one line of code in 2 places and have a more clear API interaction.

Reply:
As we discussed offline I am going to introduce additional custom field so, there won't be a need in the additional request for CI target. Therefore, all those `get*TargetMeta` functions are indeed metadata getter, not just checker.

Reply:
I buy an argument about `get*TargetMeta`, but it does not rejects a need to simplify the code.

If your forward-looking statement is valid, then the `checkAndGetTargetInfo` becomes even less intuitive, as the needed target could be fetched straight inside the `if..else` clauses (in the future), requiring less code.

Reply:
The problem I had with current code is because the `checkAndGetTargetInfo` accepts two functions as arguments, that is already mindblowing for a regular engineer.
Now, a `getTargetMeta` in a function call magically becomes a `checkTargetMeta` inside `checkAndGetTargetInfo`, while a single-line anonymous function becomes a `getTargetMeta`.
On top of that this code defines a total of 4 anonymous functions, which are only called once - another red flag.

Long story short: it took a decent amount of mental power to understand what was going on.
That would be easier to understand with more procedural and less functional programming.

I know we all shine in various high-level OP paradigms, and at the end I totally understand the implementation.
I only try to care about a less advanced reader of it, and from that perspective I see no benefit of uber-functional programming here.

Reply:
but it looks cool, isn't? :)

OK, let me try non-functional approach then.

Reply:
> but it looks cool, isn't? :)

Yes, it looks uber-cool :)

I often try functional programming to test my level of coolness, then scramble it, and commit to less shiny yet more readable procedural way.

Reply:
2a8847d